src/clp_ffi_js/ir/StreamReader.cpp

```diff
@@ -165,36 +209,37 @@ auto StreamReader::decode_range(size_t begin_idx, size_t end_idx) const -> Decod
         }
         message.append(parsed.value());
 
-        constexpr size_t cLogLevelPositionInMessages{1};
-        size_t log_level{cLogLevelNone};
-        // NOLINTNEXTLINE(readability-qualified-auto)
-        auto const log_level_name_it{std::find_if(
-                cLogLevelNames.begin() + cValidLogLevelsBeginIdx,
-                cLogLevelNames.end(),
-                [&](std::string_view level) {
-                    return message.substr(cLogLevelPositionInMessages).starts_with(level);
-                }
-        )};
-        if (log_level_name_it != cLogLevelNames.end()) {
-            log_level = std::distance(cLogLevelNames.begin(), log_level_name_it);
-        }
-
         m_ts_pattern.insert_formatted_timestamp(log_event.get_timestamp(), message);
 
         EM_ASM(
                 { Emval.toValue($0).push([UTF8ToString($1), $2, $3, $4]); },
                 results.as_handle(),
                 message.c_str(),
                 log_event.get_timestamp(),
-                log_level,
-                log_num
+                log_event.get_log_level(),
+                log_event_idx + 1
         );
-        ++log_num;
     }
 
     return DecodedResultsTsType(results);
 }
 
+void StreamReader::filter_log_events(emscripten::val const& logLevelFilter) {
+    if (logLevelFilter.isNull()) {
+        m_filtered_log_event_map.reset();
+        return;
+    }
+
+    m_filtered_log_event_map.emplace();
+    std::vector<int> filter_levels = emscripten::vecFromJSArray<int>(logLevelFilter);
+
+    for (auto const& [logEventIdx, logEvent] : std::views::enumerate(m_encoded_log_events)) {
+        if (std::ranges::find(filter_levels, logEvent.get_log_level()) != filter_levels.end()) {
+            m_filtered_log_event_map->push_back(logEventIdx);
```

Comment:
Since logEventIdx is an lvalue, copying is done even when using `emplace_back()`; also logEventIdx is as small as an int so likely there's no perf impact. With that said, do we always want to stick with `emplace_back()`?

Reply:
> With that said, do we always want to stick with emplace_back()?

Yeah, I think that's safer.